### PR TITLE
artifactregistry: implement upstream credentials

### PR DESCRIPTION
**Pull request by Subserial**
**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
artifactregistry: added `remote_repository_config.upstream_credentials` field to `google_artifact_registry_repository` resource
```

### CODE REVIEW

File: mmv1/products/artifactregistry/Repository.yaml [hunk 1]
```diff
@@ -504,7 +514,30 @@ properties:
                 description: |-
                   Specific repository from the base, e.g. `"centos/8-stream/BaseOS/x86_64/os"`
                 immutable: true
-
+      - !ruby/object:Api::Type::NestedObject
+        name: 'upstreamCredentials'
+        description: |-
+          The credentials used to access the remote repository.
+        immutable: true
```
Comment (melinath) [inline mmv1/products/artifactregistry/Repository.yaml:521]: These fields should likely all be marked sensitive:

```suggestion
        immutable: true
        sensitive: true
```

Comment (melinath) [reply mmv1/products/artifactregistry/Repository.yaml:521]: if they are sensitive, I assume that the API probably doesn't return a value for them? If that's correct, you will also need to ignore the value from the API. Unfortunately ignore_read is not currently supported on nested fields: https://github.com/hashicorp/terraform-provider-google/issues/12410 You'll need a custom flatten function similar to https://github.com/GoogleCloudPlatform/magic-modules/blob/5923d4cb878396a04bed9beaf22a8478e8b1e6a5/mmv1/products/sql/SourceRepresentationInstance.yaml#L95

File: mmv1/templates/terraform/examples/artifact_registry_repository_remote_custom.tf.erb [hunk 2]
```diff
@@ -0,0 +1,31 @@
+resource "google_secret_manager_secret" "<%= ctx[:vars]['secret_resource_id'] %>" {
+  secret_id = "<%= ctx[:vars]['secret_id'] %>"
+  replication {
+    auto {}
+  }
+}
+
+resource "google_secret_manager_secret_version" "<%= ctx[:vars]['secret_resource_id'] %>_version" {
```
Comment (melinath) [inline mmv1/templates/terraform/examples/artifact_registry_repository_remote_custom.tf.erb:8]: Tests are failing with:

```
Error: Error creating Repository: googleapi: Error 400: An error occurred while validated upstream credentials: Artifact Registry service account "service-01234567890@gcp-sa-artifactregistry.iam.gserviceaccount.com" does not have permission to access the secret version.
```

You'll need to add this permission as part of the example, similar to https://github.com/GoogleCloudPlatform/magic-modules/blob/5923d4cb878396a04bed9beaf22a8478e8b1e6a5/mmv1/templates/terraform/examples/cloudbuild_trigger_webhook_config.tf.erb#L31

**Review by melinath — CHANGES_REQUESTED** [review]
A couple changes to suggest

File: mmv1/templates/terraform/examples/artifact_registry_repository_remote_custom.tf.erb (hunk 2 above)
Comment (melinath) [reply mmv1/templates/terraform/examples/artifact_registry_repository_remote_custom.tf.erb:8]: Better example - doesn't need to use a full policy: https://github.com/GoogleCloudPlatform/magic-modules/blob/5923d4cb878396a04bed9beaf22a8478e8b1e6a5/mmv1/templates/terraform/examples/cloud_run_service_secret_environment_variables.tf.erb#L16

File: mmv1/products/artifactregistry/Repository.yaml (hunk 1 above)
Comment (melinath) [reply mmv1/products/artifactregistry/Repository.yaml:521]: If these fields are not returned by the API, you'll also need to use `ignore_read_extra` to make them not be expected during the ImportStateVerify step.

Comment (Subserial) [reply mmv1/products/artifactregistry/Repository.yaml:521]: The fields are returned by the API. The sensitive credentials are stored in the secret.

Comment (melinath) [reply mmv1/products/artifactregistry/Repository.yaml:521]: yep, I read too quickly - this is referencing a secret version for the password so all good regarding this thread, thanks!

File: mmv1/templates/terraform/examples/artifact_registry_repository_remote_custom.tf.erb (hunk 2 above)
Comment (Subserial) [reply mmv1/templates/terraform/examples/artifact_registry_repository_remote_custom.tf.erb:10]: Thanks for the reference! I tested without secret access permissions and it worked, so I've pushed the changes for testing on github.

**Review by melinath — APPROVED** [review]

**Pull request merged**